### ISSUE TITLE
Update `upgrade.txt` for 2.3.x to 3.0.x upgrade process

### ISSUE THREAD

**Issue by morgant** [opencart/opencart#5881]
Currently, the `upgrade.txt` file (and other OpenCart documentation, probably to be dealt with separately) explains how to upgrade from 1.5.x to 2.x, specifically by copying the new `upgrade/` path over the existing installation. Unfortunately, OpenCart 3.0.x has enough significant changes (naturally all the `.tpl` files that have been replaced with `.twig` files that would need to be removed, plus a number of removed JS and other files throughout the installation), that uploading the new version on top of a 2.3.x installation will not work well and introduces issues.

The `upgrade.txt` file should have a section for upgrading from 2.3.x to 3.0.x that describes not installing the new version over the old version. Also, since 3.x changes theme/modification/extension variable names (since it now requires appropriate extension type prefixes for setting names), the state of enabled extensions as well as individual settings is lost during the upgrade process, so instructions to document these settings needs to be included.

I started a thread on the OpenCart forum to document a more appropriate upgrade process: https://forum.opencart.com/viewtopic.php?f=199&t=186712

**Comment by asJulius** [comment]
I have a question about upgrading. I have modified one line in one of controllers. If I would want to upgrade OC later in the future, I would have to override that controller with the one from new OC version and redo that changed line? And if I had many files changed, then what? It looks like **OC upgrade is a NIGHTMARE**. If I have an existing application, then I should stick with that OC version it's running at?

**Comment by stalker780** [comment]
@asJulius
That is why you should use events and ocmod and not change core files directly.

**Comment by OC2PS** [comment]
Also, upgrade.txt should include instruction to upgrade from 3.x to latest 3.x

**Comment by OC2PS** [comment]
@stalker780 
> Some people prefer not to upgrade their shops for decades
Yeah, and @danielkerr sometimes gets angry at people upset about the lack of clear upgrade paths, saying "if your store is working fine, don't upgrade"
I understand that view.
However, it's not 100% honest either. The older versions are not being maintained, so security vulnerabilities that do get discovered later will never be patched (in the old versions). So there's a security risk. Yeah, when the old version was created, it may have been bulletproof, but hackers come up with new ways and capabilities all the time, and so the possibility of new vulnerabilities in old versions is there.
On top of that, sometimes, just the supporting infrastructure may render the old versions obsolete. e.g. Now that PHP 7 is over 2 years old, some hosts have started migrating to it. This may create issues with older versions.
So, technically, while if your store is working fine, don't upgrade, the fact is that your store may not stay working fine if you don't upgrade.

**Comment by Shifterovich** [comment]
@OC2PS 

> The older versions are not being maintained, so security vulnerabilities that do get discovered later will never be patched (in the old versions).

When using the latest release, you can't even pull changes from GitHub, since they often change how things work, which breaks for example extensions. Maybe it'd be a good idea to have branches for every major release with patches that won't break anything? Security and small bugs.

**Comment by danielkerr** [comment]
@OC2PS

`However, it's not 100% honest either. The older versions are not being maintained, so security vulnerabilities that do get discovered later will never be patched (in the old versions). So there's a security risk`

its not own to me to maintain older versions! its down to the site builders to keep up to date with security fixes added to opencart! dont come and post lies!

**Comment by OC2PS** [comment]
@danielkerr Not sure why I angered you. I didn't say that it's your responsibility to update everyone's website. What I said was that people need a update path. In the past you have sometimes said "don't upgrade if your site is working ok"...I was merely saying that's not a good reason for not having an update path...people do need to update because they should be updating to newest versions because you and your team are patching bugs and security issues all the time.

**Comment by Shifterovich** [comment]
@danielkerr Having an update function that would pull very small fixes like I mentioned from some git branch is something that could be done and would make this much easier for the site builders.

> its down to the site builders to keep up to date with security fixes added to opencart

Often very hard when you can't just click "Update" and can't code.

**Comment by danielkerr** [comment]
Well that's how scripts work. They are not packaged apps

**Comment by Shifterovich** [comment]
I don't see why you couldn't download and apply small fixes using PHP.

**Comment by Shifterovich** [comment]
Could be something like git commits, or separate patches (or many, packed together) as extensions. Former would be "cleaner", though, I think.

**Comment by danielkerr** [comment]
The issue is people modifing code or templates with missing vars, language,
mods etc..

**Comment by Shifterovich** [comment]
You mean merge conflicts when you'd pull the patches? Thought about that, and if there were any changes made directly to the core files, the user could be presented with some options to resolve the conflict:

1. Keep the old version
2. Keep the new version
3. Manually merge the two versions

**Comment by xdjxdaveyx** [comment]
I think a good idea to make upgrades easier would just to have an old database upgrader... your could run the program on your existing sql to create an sql compatible with the latest fresh install of the new opencart and just load your updated sql into it. The biggest obstacle i've found is with the database - if we can upgrade this with a separate program there is less chance of data loss and crashed sites due to people not jumping straight to a fresh install?

**Comment by xdjxdaveyx** [comment]
..or perhaps just an option to load old database structures into new versions of opencart which can then convert the architecture - but this is more work for OC designers :(
